File: tools/sp-debug/src/sp_debug/dml_preview.py
"""Convert DML statements into safe SELECT previews for debug harness scripts."""

from __future__ import annotations

import re

CLAUSE_FROM = re.compile(r"\bFROM\b", re.IGNORECASE)
CLAUSE_WHERE = re.compile(r"\bWHERE\b", re.IGNORECASE)
CLAUSE_SET = re.compile(r"\bSET\b", re.IGNORECASE)
INSERT_INTO = re.compile(r"^\s*INSERT\s+INTO\s+(\S+)", re.IGNORECASE)
DELETE_FROM = re.compile(
    r"^\s*DELETE\s+FROM\s+(\S+)(?:\s+WHERE\s+(.+))?\s*;?\s*$",
    re.IGNORECASE | re.DOTALL,
)


def _extract_paren_content(text: str, open_index: int) -> tuple[str, int] | None:
    if open_index >= len(text) or text[open_index] != "(":
        return None
    depth = 0
    for i in range(open_index, len(text)):
        ch = text[i]
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
            if depth == 0:
                return text[open_index + 1 : i], i + 1
    return None


def _block_sql(block_lines: list[str]) -> str:
    parts = [ln.strip() for ln in block_lines if ln.strip()]
    return re.sub(r"\s+", " ", " ".join(parts)).strip()


def _find_clause(text: str, pattern: re.Pattern[str]) -> int | None:
    match = pattern.search(text)
    return match.start() if match else None


def _split_expressions(text: str) -> list[str]:
    parts: list[str] = []
    depth = 0
    current: list[str] = []
    for ch in text:
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
        elif ch == "," and depth == 0:
            part = "".join(current).strip()
            if part:
                parts.append(part)
            current = []
            continue
        current.append(ch)
    tail = "".join(current).strip()
    if tail:
        parts.append(tail)
    return parts


def _parse_assignments(set_clause: str) -> list[tuple[str, str]]:
    assignments: list[tuple[str, str]] = []
    for part in _split_expressions(set_clause):
        if "=" not in part:
            continue
        lhs, rhs = part.split("=", 1)
        assignments.append((lhs.strip(), rhs.strip()))
    return assignments


def _column_alias(name: str) -> str:
    base = name.split(".")[-1].strip("[]")
    return f"[{base}]"


_BARE_VAR = re.compile(r"^@\w+$", re.IGNORECASE)
_QUOTED_LITERAL = re.compile(r"^(N?'([^']|'')*'|\d+(\.\d+)?)$", re.IGNORECASE)


def _lhs_column_name(lhs: str) -> str:
    return lhs.split(".")[-1].strip("[]")


def _is_calculation(expr: str) -> bool:
    """True when RHS is an expression, not a bare variable or simple literal."""
    text = expr.strip()
    if _BARE_VAR.match(text):
        return False
    if _QUOTED_LITERAL.match(text):
        return False
    if re.search(r"[\+\-\*/%]|^\w+\(", text, re.IGNORECASE):
        return True
    return False


def _preview_column_alias(lhs: str, rhs: str) -> str:
    """Choose preview column alias based on SET/VALUES expression shape."""
    rhs = rhs.strip()
    if _BARE_VAR.match(rhs):
        return f"[{rhs}]"
    col = _lhs_column_name(lhs)
    if _is_calculation(rhs):
        return f"[calculated-{col}]"
    return f"[{col}]"


def _parse_update(sql: str) -> dict[str, str | list[tuple[str, str]]] | None:
    text = sql.strip().rstrip(";")
    if not re.match(r"UPDATE\b", text, re.I):
        return None

    rest = re.sub(r"^UPDATE\s+", "", text, flags=re.I).strip()
    set_match = CLAUSE_SET.search(rest)
    if not set_match:
        return None

    target = rest[: set_match.start()].strip()
    after_set = rest[set_match.end() :].strip()
    from_pos = _find_clause(after_set, CLAUSE_FROM)
    where_pos = _find_clause(after_set, CLAUSE_WHERE)
    stops = [pos for pos in (from_pos, where_pos) if pos is not None]
    assign_end = min(stops) if stops else len(after_set)

    assignments = _parse_assignments(after_set[:assign_end])
    if not assignments:
        return None

    from_clause = ""
    if from_pos is not None:
        from_end = where_pos if where_pos is not None and where_pos > from_pos else len(after_set)
        from_clause = after_set[from_pos:from_end].strip()

    where_clause = ""
    if where_pos is not None:
        where_clause = after_set[where_pos:].strip()

    return {
        "kind": "UPDATE",
        "target": target,
        "assignments": assignments,
        "from_clause": from_clause,
        "where_clause": where_clause,
    }


def _parse_insert(sql: str) -> dict[str, str | list[str]] | None:
    one_line = re.sub(r"\s+", " ", sql.strip().rstrip(";"))
    head = INSERT_INTO.match(one_line)
    if not head:
        return None

    table = head.group(1)
    pos = head.end()
    while pos < len(one_line) and one_line[pos].isspace():
        pos += 1
    columns: list[str] = []

    if pos < len(one_line) and one_line[pos] == "(":
        col_content, pos = _extract_paren_content(one_line, pos) or ("", pos)
        columns = [c.strip() for c in col_content.split(",") if c.strip()]

    values_match = re.search(r"\bVALUES\b", one_line[pos:], re.I)
    if not values_match:
        return None
    pos += values_match.end()
    while pos < len(one_line) and one_line[pos].isspace():
        pos += 1

    values_content, _end = _extract_paren_content(one_line, pos) or ("", pos)
    values = _split_expressions(values_content)
    if not values:
        return None
    if columns and len(columns) != len(values):
        return None
    if not columns:
        columns = [f"col{i + 1}" for i in range(len(values))]

    return {
        "kind": "INSERT",
        "target": table,
        "columns": columns,
        "values": values,
    }


def _parse_delete(sql: str) -> dict[str, str] | None:
    one_line = re.sub(r"\s+", " ", sql.strip().rstrip(";"))
    match = DELETE_FROM.match(one_line)
    if not match:
        return None
    return {
        "kind": "DELETE",
        "target": match.group(1),
        "where_clause": (match.group(2) or "").strip(),
    }


def _format_multiline(indent: str, sql: str) -> list[str]:
    return [f"{indent}{line}" if line else "" for line in sql.splitlines()]


def build_dml_preview(block_lines: list[str], indent: str) -> list[str] | None:
    """Return SELECT preview lines replacing a DML block, or None if unsupported."""
    sql = _block_sql(block_lines)
    first = block_lines[0].strip().split()[0].upper()

    if first == "UPDATE":
        parsed = _parse_update(sql)
        if parsed is None:
            return None
        target = str(parsed["target"])
        assignments = parsed["assignments"]
        assert isinstance(assignments, list)

        select_cols = [f"N'UPDATE to table {target}' AS [DBG_Action]"]
        for lhs, rhs in assignments:
            select_cols.append(f"{rhs} AS {_preview_column_alias(str(lhs), str(rhs))}")

        from_clause = str(parsed.get("from_clause") or "")
        if from_clause:
            from_sql = from_clause
        else:
            from_sql = f"FROM {target}"

        where_clause = str(parsed.get("where_clause") or "").strip()
        preview = (
            f"SELECT {', '.join(select_cols)}\n"
            f"{from_sql}"
            + (f"\n{where_clause}" if where_clause else "")
            + ";"
        )
        return [
            f"{indent}-- [DBG-PREVIEW] Would have executed:",
            *_format_multiline(indent, preview),
        ]

    if first == "INSERT":
        parsed = _parse_insert(sql)
        if parsed is None:
            return None
        target = str(parsed["target"])
        columns = parsed["columns"]
        values = parsed["values"]
        assert isinstance(columns, list) and isinstance(values, list)

        select_cols = [f"N'INSERT to table {target}' AS [DBG_Action]"]
        for col, val in zip(columns, values, strict=True):
            select_cols.append(f"{val} AS {_preview_column_alias(str(col), str(val))}")

        preview = f"SELECT {', '.join(select_cols)};"
        return [
            f"{indent}-- [DBG-PREVIEW] Would have executed:",
            *_format_multiline(indent, preview),
        ]

    if first == "DELETE":
        parsed = _parse_delete(sql)
        if parsed is None:
            return None
        target = str(parsed["target"])
        where_clause = str(parsed.get("where_clause") or "").strip()
        select_cols = [f"N'DELETE from table {target}' AS [DBG_Action]", "*"]
        preview = (
            f"SELECT {', '.join(select_cols)}\n"
            f"FROM {target}"
            + (f"\nWHERE {where_clause}" if where_clause else "")
            + ";"
        )
        return [
            f"{indent}-- [DBG-PREVIEW] Would have executed:",
            *_format_multiline(indent, preview),
        ]

    return None

File: src/engine/dmlPreview.ts
import {
  BARE_VAR,
  CALCULATION_PATTERN,
  CLAUSE_FROM,
  CLAUSE_SET,
  CLAUSE_WHERE,
  DELETE_FROM_LINE,
  INSERT_INTO_LINE,
  QUOTED_LITERAL,
  UPDATE_TARGET,
} from "./constants";

function extractParenContent(
  text: string,
  openIndex: number
): [string, number] | null {
  if (openIndex >= text.length || text[openIndex] !== "(") {
    return null;
  }
  let depth = 0;
  for (let i = openIndex; i < text.length; i++) {
    const ch = text[i];
    if (ch === "(") {
      depth += 1;
    } else if (ch === ")") {
      depth -= 1;
      if (depth === 0) {
        return [text.slice(openIndex + 1, i), i + 1];
      }
    }
  }
  return null;
}

function blockSql(blockLines: string[]): string {
  const parts = blockLines.map((ln) => ln.trim()).filter(Boolean);
  return parts.join(" ").replace(/\s+/g, " ").trim();
}

function findClause(text: string, pattern: RegExp): number | null {
  const match = pattern.exec(text);
  return match ? match.index : null;
}

function splitExpressions(text: string): string[] {
  const parts: string[] = [];
  let depth = 0;
  let current = "";
  for (const ch of text) {
    if (ch === "(") {
      depth += 1;
    } else if (ch === ")") {
      depth -= 1;
    } else if (ch === "," && depth === 0) {
      const part = current.trim();
      if (part) {
        parts.push(part);
      }
      current = "";
      continue;
    }
    current += ch;
  }
  const tail = current.trim();
  if (tail) {
    parts.push(tail);
  }
  return parts;
}

function parseAssignments(setClause: string): Array<[string, string]> {
  const assignments: Array<[string, string]> = [];
  for (const part of splitExpressions(setClause)) {
    if (!part.includes("=")) {
      continue;
    }
    const eq = part.indexOf("=");
    assignments.push([part.slice(0, eq).trim(), part.slice(eq + 1).trim()]);
  }
  return assignments;
}

function lhsColumnName(lhs: string): string {
  return lhs.split(".").pop()!.replace(/^\[|\]$/g, "");
}

function isCalculation(expr: string): boolean {
  const text = expr.trim();
  if (BARE_VAR.test(text)) {
    return false;
  }
  if (QUOTED_LITERAL.test(text)) {
    return false;
  }
  return CALCULATION_PATTERN.test(text);
}

function previewColumnAlias(lhs: string, rhs: string): string {
  rhs = rhs.trim();
  if (BARE_VAR.test(rhs)) {
    return `[${rhs}]`;
  }
  const col = lhsColumnName(lhs);
  if (isCalculation(rhs)) {
    return `[calculated-${col}]`;
  }
  return `[${col}]`;
}

function parseUpdate(sql: string): {
  kind: string;
  target: string;
  assignments: Array<[string, string]>;
  fromClause: string;
  whereClause: string;
} | null {
  const text = sql.trim().replace(/;+$/, "");
  if (!UPDATE_TARGET.test(text)) {
    return null;
  }
  const rest = text.replace(/^UPDATE\s+/i, "").trim();
  const setMatch = CLAUSE_SET.exec(rest);
  if (!setMatch) {
    return null;
  }
  const target = rest.slice(0, setMatch.index).trim();
  const afterSet = rest.slice(setMatch.index + setMatch[0].length).trim();
  const fromPos = findClause(afterSet, CLAUSE_FROM);
  const wherePos = findClause(afterSet, CLAUSE_WHERE);
  const stops = [fromPos, wherePos].filter((p): p is number => p !== null);
  const assignEnd = stops.length ? Math.min(...stops) : afterSet.length;
  const assignments = parseAssignments(afterSet.slice(0, assignEnd));
  if (!assignments.length) {
    return null;
  }

  let fromClause = "";
  if (fromPos !== null) {
    const fromEnd =
      wherePos !== null && wherePos > fromPos ? wherePos : afterSet.length;
    fromClause = afterSet.slice(fromPos, fromEnd).trim();
  }

  let whereClause = "";
  if (wherePos !== null) {
    whereClause = afterSet.slice(wherePos).trim();
  }

  return { kind: "UPDATE", target, assignments, fromClause, whereClause };
}

function parseInsert(sql: string): {
  kind: string;
  target: string;
  columns: string[];
  values: string[];
} | null {
  const oneLine = sql.trim().replace(/;+$/, "").replace(/\s+/g, " ");
  const head = INSERT_INTO_LINE.exec(oneLine);
  if (!head) {
    return null;
  }
  const table = head[1];
  let pos = head[0].length;
  while (pos < oneLine.length && /\s/.test(oneLine[pos])) {
    pos += 1;
  }
  let columns: string[] = [];
  if (pos < oneLine.length && oneLine[pos] === "(") {
    const extracted = extractParenContent(oneLine, pos);
    if (extracted) {
      const [colContent, nextPos] = extracted;
      columns = colContent
        .split(",")
        .map((c) => c.trim())
        .filter(Boolean);
      pos = nextPos;
    }
  }

  const valuesMatch = /\bVALUES\b/i.exec(oneLine.slice(pos));
  if (!valuesMatch) {
    return null;
  }
  pos += valuesMatch.index! + valuesMatch[0].length;
  while (pos < oneLine.length && /\s/.test(oneLine[pos])) {
    pos += 1;
  }

  const valuesExtracted = extractParenContent(oneLine, pos);
  if (!valuesExtracted) {
    return null;
  }
  const values = splitExpressions(valuesExtracted[0]);
  if (!values.length) {
    return null;
  }
  if (columns.length && columns.length !== values.length) {
    return null;
  }
  if (!columns.length) {
    columns = values.map((_, i) => `col${i + 1}`);
  }

  return { kind: "INSERT", target: table, columns, values };
}

function parseDelete(sql: string): {
  kind: string;
  target: string;
  whereClause: string;
} | null {
  const oneLine = sql.trim().replace(/;+$/, "").replace(/\s+/g, " ");
  const match = DELETE_FROM_LINE.exec(oneLine);
  if (!match) {
    return null;
  }
  return {
    kind: "DELETE",
    target: match[1],
    whereClause: (match[2] || "").trim(),
  };
}

function formatMultiline(indent: string, sql: string): string[] {
  return sql.split(/\r?\n/).map((line) => (line ? `${indent}${line}` : ""));
}

export function buildDmlPreview(
  blockLines: string[],
  indent: string
): string[] | null {
  const sql = blockSql(blockLines);
  const first = blockLines[0].trim().split(/\s+/)[0].toUpperCase();

  if (first === "UPDATE") {
    const parsed = parseUpdate(sql);
    if (!parsed) {
      return null;
    }
    const selectCols = [`N'UPDATE to table ${parsed.target}' AS [DBG_Action]`];
    for (const [lhs, rhs] of parsed.assignments) {
      selectCols.push(`${rhs} AS ${previewColumnAlias(lhs, rhs)}`);
    }
    const fromSql = parsed.fromClause || `FROM ${parsed.target}`;
    const whereClause = parsed.whereClause.trim();
    const preview =
      `SELECT ${selectCols.join(", ")}\n` +
      fromSql +
      (whereClause ? `\n${whereClause}` : "") +
      ";";
    return [
      `${indent}-- [DBG-PREVIEW] Would have executed:`,
      ...formatMultiline(indent, preview),
    ];
  }

  if (first === "INSERT") {
    const parsed = parseInsert(sql);
    if (!parsed) {
      return null;
    }
    const selectCols = [`N'INSERT to table ${parsed.target}' AS [DBG_Action]`];
    for (let i = 0; i < parsed.columns.length; i++) {
      selectCols.push(
        `${parsed.values[i]} AS ${previewColumnAlias(parsed.columns[i], parsed.values[i])}`
      );
    }
    const preview = `SELECT ${selectCols.join(", ")};`;
    return [
      `${indent}-- [DBG-PREVIEW] Would have executed:`,
      ...formatMultiline(indent, preview),
    ];
  }

  if (first === "DELETE") {
    const parsed = parseDelete(sql);
    if (!parsed) {
      return null;
    }
    const selectCols = [
      `N'DELETE from table ${parsed.target}' AS [DBG_Action]`,
      "*",
    ];
    const preview =
      `SELECT ${selectCols.join(", ")}\n` +
      `FROM ${parsed.target}` +
      (parsed.whereClause ? `\nWHERE ${parsed.whereClause}` : "") +
      ";";
    return [
      `${indent}-- [DBG-PREVIEW] Would have executed:`,
      ...formatMultiline(indent, preview),
    ];
  }

  return null;
}
